builtin/credential/ldap/backend.go

```diff
@@ -101,106 +102,191 @@ func (b *backend) Login(req *logical.Request, username string, password string)
-			return nil, logical.ErrorResponse(fmt.Sprintf("LDAP search failed: %v", err)), nil
+			return userDN, fmt.Errorf("LDAP search failed for detecting user: %v", err)
 		}
-		for _, e := range sresult.Entries {
-			userdn = e.DN
+		for _, e := range result.Entries {
+			userDN = e.DN
 		}
 	} else {
-		userdn = binddn
+		userDN = bindDN
 	}
 
-	var allgroups []string
-	var policies []string
-	resp := &logical.Response{
-		Data: map[string]interface{}{},
+	return userDN, nil
+}
+
+func getLdapGroups(cfg *ConfigEntry, c *ldap.Conn, userDN string, username string) ([]string, error) {
+	// retrieve the groups in a string/bool map as a structure to avoid duplicates inside
+	ldapMap := make(map[string]bool)
+	// Fetch the optional memberOf property values on the user object
+	// This is the most common method used in Active Directory setup to retrieve the groups
+	result, err := c.Search(&ldap.SearchRequest{
+		BaseDN: userDN,
+		Scope:  0,        // base scope to fetch only the userDN
+		Filter: "(cn=*)", // bogus filter, required to fetch the CN from userDN
+		Attributes: []string{
+			"memberOf",
+		},
+	})
+	// this check remains in case something happens with the ldap query or connection
+	if err != nil {
+		return nil, fmt.Errorf("LDAP fetch of distinguishedName=%s failed: %v", userDN, err)
 	}
+	// if there are more than one entry, we consider the results irrelevant and ignore them
+	if len(result.Entries) == 1 {
+		for _, attr := range result.Entries[0].Attributes {
+			// Find the groups the user is member of from the 'memberOf' attribute extracting the CN
+			if attr.Name == "memberOf" {
+				for _, value := range attr.Values {
+					memberOfDN, err := ldap.ParseDN(value)
+					if err != nil || len(memberOfDN.RDNs) == 0 {
+						continue
+					}
 
-	// Fetch custom (local) groups the user has been added to
-	user, err := b.User(req.Storage, username)
-	if err == nil && user != nil {
-		allgroups = append(allgroups, user.Groups...)
+					for _, rdn := range memberOfDN.RDNs {
+						for _, rdnTypeAndValue := range rdn.Attributes {
+							if strings.EqualFold(rdnTypeAndValue.Type, "CN") {
+								ldapMap[rdnTypeAndValue.Value] = true
+							}
+						}
+					}
+				}
+			}
+		}
 	}
 
+	// Find groups by searching in groupDN for any of the memberUid, member or uniqueMember attributes
+	// and retrieving the CN in the DN result
 	if cfg.GroupDN != "" {
-		// Enumerate all groups the user is member of. The search filter should
-		// work with both openldap and MS AD standard schemas.
-		sresult, err := c.Search(&ldap.SearchRequest{
+		result, err := c.Search(&ldap.SearchRequest{
 			BaseDN: cfg.GroupDN,
 			Scope:  2, // subtree
-			Filter: fmt.Sprintf("(|(memberUid=%s)(member=%s)(uniqueMember=%s))", ldap.EscapeFilter(username), ldap.EscapeFilter(userdn), ldap.EscapeFilter(userdn)),
+			Filter: fmt.Sprintf("(|(memberUid=%s)(member=%s)(uniqueMember=%s))", ldap.EscapeFilter(username), ldap.EscapeFilter(userDN), ldap.EscapeFilter(userDN)),
```

Comment:
@steve-jansen negative. According to [Microsoft's documentation](https://msdn.microsoft.com/en-us/library/ms677943.aspx):
>Be aware that this attribute lists the groups that contain the user in their member attribute—it does not contain the recursive list of nested predecessors. For example, if user O is a member of group C and group B and group B were nested in group A, the memberOf attribute of user O would list group C and group B, but not group A.

I can also confirm that querying the `memberOf` attribute for my user returns _20_ results - my direct group memberships, while using the `LDAP_MATCHING_RULE_IN_CHAIN` I referenced earlier returns **120** groups - direct + transitive membership.